Add CGI functionality

cpq opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cesanta/fossa/251)
<!-- Reviewable:end -->

alashkin commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cesanta/fossa/251)

Reviewed files:
- examples/restful_server/Makefile @ r1
- examples/restful_server/restful_server.c @ r1
- fossa.c @ r1
- fossa.h @ r1
- src/common.h @ r1
- src/http.c @ r1
- src/http.h @ r1
- test/data/cgi/index.cgi @ r1
- test/unit_test.c @ r1





---

<sup>**[src/http.c, line 530 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlojAqPNFY_BueFnSK2)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L530)):</sup>
`CGI!!`???

---

<sup>**[src/http.c, line 1687 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlokTYtpj7PJI14hBAD)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1687)):</sup>
Is it correct to zero the _last_ symbol in path?
Should not it be `path[path_len - n + <sprintf result>]`?

---

<sup>**[src/http.c, line 1753 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlolFRBTEFmj8QRNd1c)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1753)):</sup>
May be `return select(sock + 1, for_read ? &set : 0, for_read ? 0 : &set, 0, 0) != SOCKET_ERROR;` will be better?

---

<sup>**[src/http.c, line 1881 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-Jlom9f2xaFAWbV29-vc)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1881)):</sup>
Dead code?

---

<sup>**[src/http.c, line 1935 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlomteiqOr7ymeM1Y4O)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1935)):</sup>
if `block->len == sizeof(block->buf)` space is negative and protected by assert only.
Is it impossible situation and `assert` (in debug mode only) is OK here?

---

<sup>**[src/http.c, line 1936 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlomcZkvRTgUvTOxCSh)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1936)):</sup>
But you didn't want to have `assert` in library code :)

---

<sup>**[src/http.h, line 20 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JloegFf7tWrfAXt-k7M)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.h#L20)):</sup>
Shouldn't we surround all size-define with
`#ifndef`? Like
```
#ifndef NS_CGI_ENVIRONMENT_SIZE
#define NS_CGI_ENVIRONMENT_SIZE 8192
#endif
```
8Kb is very big buffer, it won't work on embedded.
It could be separated issue with separated PR.

Macro redefinition is warning by default, but I compile AVR with -Werror (coz some of warnings are very important).

---


<!-- Sent from Reviewable.io -->

cpq commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cesanta/fossa/251)







---

<sup>**[src/http.c, line 530 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlojAqPNFY_BueFnSK2)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L530)):</sup>
Fixed :-)

---

<sup>**[src/http.c, line 1687 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlokTYtpj7PJI14hBAD)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1687)):</sup>
It's just a protection from overflow.
snprintf, by standard, can return a result that is larger than buffer length, in which case it won't 0-terminate the string. We protect against it.
Normally, snprintf will 0-terminate the string somewhere in the middle of a buffer, and we also add a protective 0-terminator at the end of a buffer.

---

<sup>**[src/http.c, line 1753 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlolFRBTEFmj8QRNd1c)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1753)):</sup>
Done.

---

<sup>**[src/http.c, line 1881 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-Jlom9f2xaFAWbV29-vc)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1881)):</sup>
Added comment.

---

<sup>**[src/http.c, line 1935 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlomteiqOr7ymeM1Y4O)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1935)):</sup>
Removed asserts

---

<sup>**[src/http.c, line 1936 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlomcZkvRTgUvTOxCSh)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1936)):</sup>
Yeah I don't! Removed.

---

<sup>**[src/http.h, line 20 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JloegFf7tWrfAXt-k7M)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.h#L20)):</sup>
Good point.
Wrapped all defines in ifndef.

---


<!-- Sent from Reviewable.io -->

alashkin commented:
Reviewed files:
- fossa.c @ r2
- fossa.h @ r2
- src/http.c @ r2
- src/http.h @ r2

---

<sup>**[src/http.c, line 1687 \[r1\]](https://reviewable.io:443/reviews/cesanta/fossa/251#-JlokTYtpj7PJI14hBAD)** ([raw file](https://github.com/cesanta/fossa/blob/0791bb1dd35ec564ef759b09c7cd3c9412c334c6/src/http.c#L1687)):</sup>
Oh, stupid me )

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cesanta/fossa/251)
<!-- Sent from Reviewable.io -->
